recipe/meta.yaml

```diff
@@ -1,17 +1,18 @@
 {% set version = "6.4" %}
+{% set patch = "20240210" %}
 
 package:
   name: ncurses
-  version: {{ version }}
+  version: {{ version }}_{{ patch }}
 
 source:
-  fn: ncurses-{{ version }}.tar.gz
-  url: http://ftp.gnu.org/pub/gnu/ncurses/ncurses-{{ version }}.tar.gz
-  sha256: 6931283d9ac87c5073f30b6290c4c75f21632bb4fc3603ac8100812bed248159
+  fn: ncurses-{{ version }}-{{ patch }}.tgz
+  url: http://invisible-mirror.net/archives/ncurses/current/ncurses-{{ version }}-{{ patch }}.tgz
```

Comment:
Sorry for my ignorance but source changes on important libraries are a bit scary. Do you have any background info on what/how/why we need to change the source here?

Reply:
The main problem is that the original sources https://invisible-island.net/ncurses/ncurses.faq.html#latest_version have not been supported for a long time and the last release was in 2022, and currently only patches are being released https://invisible-island.net/archives/ncurses/6.4/. But there are alternative sources, such as https://github.com/ThomasDickey/ncurses-snapshots, which are supported (ThomasDickey is the maintainer of ncurses) and which can be used to build the component without vulnerabilities. After all, version 6.4_20221231 contains quite a lot of CVE that have been fixed in latest versions. Speaking of these sources (http://invisible-mirror.net/archives/ncurses/current), I don't know much about them, but maybe @chribl can tell us why he suggests using exactly them

Reply:
Thanks! That helps a lot!! @isuruf and/or @h-vetinari are usually more up-to-date than myself with these to judge how urgent we need to merge this update. I'll merge this in a few days if they don't comment until then.

Reply:
Could we please use this instead?

```suggestion
  url: https://ftp.gnu.org/gnu/ncurses/ncurses-{{ version }}.tar.gz
```

Reply:
Generally, CVEs create a lot of sad (or angry) faces in environments where automated tooling enforces their absence. It's IMO in the interest of conda-forge to fix these things quickly.

That said, I don't know this specifc case, though mainly it appears we just need to use up-to-date sources. Pointing towards the upstream canonical repo is usually good enough for the bot to pick up updates. Fine by me to use the gnu server, or an up-to-date mirror on GitHub/gitlab (if there is one; not 100% sure our bot detects updates on the FTP server)

Reply:
I don't know this specific case either. However:

* The PR is to use the 6.4-20240210 patch snapshot rather than the 6.4 release (which is from late 2022).
* ftp.gnu.org does not contain the patch snapshot — the most recent tarball there is the 6.4 release, which is already the existing conda-forge package.
* The ThomasDickey/ncurses-snapshots GitHub repository almost certainly contains the same code as invisible-mirror.net, which is Thomas Dickey's web site. Indeed `https://www.gnu.org/software/ncurses/` redirects to invisible-island.net, which apparently invisible-mirror.net mirrors (and appears to also be controlled by TD).

Reply:
Sorry for causing so much confusion, I should have added more description in the beginning.. 
@jmarshall Already did the anaylsis in his comment, but for completeness I add my view when creating this PR (that was mostly about discussing the possibility of an ncurses update).

* ncurses does not have a official release since a while and an open CVE in the last release
* Thomas Dickey is releasing the current development snapshots after patches are applied (e.g the lastest from Sunday https://lists.gnu.org/archive/html/bug-ncurses/2024-03/msg00002.html )
* Also Linux distributions did update to user later ncurses versions to fix the CVE ( see e.g. Arch Linuxs patch here https://gitlab.archlinux.org/archlinux/packaging/packages/ncurses/-/commit/3c2606603aa4a5a3b2d29e560a1bc14986153f49 )
* Also from the GNU mailing list for ncurses and the mail from Thomas Dickey the invisible-island.net and invisible-mirror.net are official pages from Thomas Dickey AFAIK I was using the similar link to the Arch recipe

Open questions from my side about this PR:
* Is it acceptable to use a "beta" package for ncurses to mitigate the open CVE in nucrses?
* If so what is the best "beta" version to pick? (I just used the latest as example) Another approach could be e.g. the same version as Arch Linux as it is apparently already tested there
* When checking this I noticed that the links used in this PR are not online very long (that was also the reason to use the latest), e.g. right now the last patch release available on this link is from October 2023. So if we do use this approach at some point the recipe might be broken as the URL will be offline (approx. 5-6 month after the version was published)

So for the last point some other way could be to use the github repository @jmarshall  mentioned. If this is a legit source this might be the most flexible and future proof solution IMO.

Sorry again for the confusion and I am happy to do any recommended changes to this PR, the main purpose was to fix the CVE and start the communication, I am also pretty new to the conda-forge environment and did not knew where to best post this as according to this comment https://github.com/conda-forge/ncurses-feedstock/pull/71#issuecomment-1579262566 there are not many active maintainers of this recipe.

Reply:
Then my suggestion is to use these sources https://github.com/ThomasDickey/ncurses-snapshots by referring to the latest available tag, this will be the safest way

Reply:
> Then my suggestion is to use these sources https://github.com/ThomasDickey/ncurses-snapshots by referring to the latest available tag, this will be the safest way

I added a commit to change the recipe to retrieve the source from this GitHub mirror. Also I changed it a bit that it can handle both "full" versions and patched versions from GitHub and as a package version.

Reply:
For anyone wondering, Thomas Dickey is the principal author of ncurses for the last ~25 years, see this long historical [overview](https://invisible-island.net/ncurses/ncurses.faq.html). His snapshots are about as official as it gets, certainly more than the GNU sources. They're also [updated](https://github.com/ThomasDickey/ncurses-snapshots/tags) very regularly, basically every week.

Reply:
What does Debian do? Do they have patches that we can apply?

Reply:
It looks like they're still using 6.4_20221231 for the stable debian  version (although I'm not sure). One of the latest versions of ncurses in debian was 6.4_20240113-1, but available for Sid